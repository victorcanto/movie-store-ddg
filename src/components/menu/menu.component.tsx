import React, { useEffect, useState } from 'react';
import { Box, Drawer as MuiDrawer, useTheme } from '@mui/material';

import {
	selectIsCartOpen,
	selectIsFavoritesOpen,
	setIsCartOpen,
	setIsFavoritesOpen,
	useAppDispatch,
	useAppSelector,
} from '../../store';
import { ListItems } from '../list-items/list-items.component';
import { CartTotal } from '../cart-total/cart-total.component';

export function Menu() {
	const theme = useTheme();
	const dispatch = useAppDispatch();
	const isCartOpen = useAppSelector(selectIsCartOpen);
	const isFavoriteOpen = useAppSelector(selectIsFavoritesOpen);

	const [isOpen, setIsOpen] = useState<boolean>(false);
	const [title, setTitle] = useState<string>('');
	const [emptyMessage, setEmptyMessage] = useState<string>('');

	const closeMenu = (): void => {
		setIsOpen(false);
	};

	const onCloseMenuHandler = (): void => {
		dispatch(setIsCartOpen(false));
		dispatch(setIsFavoritesOpen(false));
		closeMenu();
	};
	const openMenu = (): void => setIsOpen(true);

	const cartTitle = (): void => setTitle('Meu Carrinho');
	const favoritesTitle = (): void => setTitle('Meus Favoritos');

	const cartEmptyMessage = (): void =>
		setEmptyMessage('Nenhum filme foi adicionado.');
	const favoritesEmptyMessage = (): void =>
		setEmptyMessage('Nenhum filme favoritado.');

	useEffect(() => {
		if (!isCartOpen && !isFavoriteOpen) {
			closeMenu();
		} else if (isCartOpen && !isFavoriteOpen) {
			openMenu();
			cartTitle();
			cartEmptyMessage();
		} else if (!isCartOpen && isFavoriteOpen) {
			openMenu();
			favoritesTitle();
			favoritesEmptyMessage();
		}
	}, [isCartOpen, isFavoriteOpen]);

	return (
		<MuiDrawer
			data-testid='menu-drawer'
			transitionDuration={theme.transitions.duration.short}
			variant='persistent'
			PaperProps={{
				sx: {
					p: '1rem',
					height: '100%',
					bottom: 0,
					top: '65px',
				},
			}}
			anchor='right'
			open={isOpen}
		>
			<Box
				sx={{
					display: 'flex',
					flexDirection: 'column',
					justifyContent: 'space-between',
					width: { xs: '80vw', md: '30vw' },
					mb: '12rem',
				}}
				role='presentation'
			>
				<Box
					sx={{
						display: 'flex',
						flexDirection: 'column',
						justifyContent: 'space-between',
					}}
				>
					<ListItems
						onCloseMenuHandler={onCloseMenuHandler}
						title={title}
						emptyMessage={emptyMessage}
					/>
				</Box>
				<Box
					sx={{
						backgroundColor: theme.palette.common.white,
						position: 'fixed',
						bottom: 0,
						width: { xs: '80vw', md: '30vw' },
					}}
				>
					{isCartOpen && (
						<CartTotal onCloseMenuHandler={onCloseMenuHandler} />
					)}
				</Box>
			</Box>
		</MuiDrawer>
	);
}
