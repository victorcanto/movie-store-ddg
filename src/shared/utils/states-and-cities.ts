import statesAndCities from '../data/estados-cidades.json';

export const getBrazilianStates = () =>
	statesAndCities.estados.map(({ nome }) => nome);

export const getBrazilianCitiesByState = (stateName: string) =>
	statesAndCities.estados.find(({ nome }) => nome === stateName)?.cidades;
