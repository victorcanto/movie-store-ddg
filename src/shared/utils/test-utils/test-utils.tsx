import React, { PropsWithChildren } from 'react';
import { render as rtlRender } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { setupStore, AppStore, RootState } from '../../../store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: AppStore;
}

export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {},
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {},
) {
	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return (
			<Provider store={store}>
				<BrowserRouter>{children}</BrowserRouter>
			</Provider>
		);
	}
	return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';

export { renderWithProviders as render };
