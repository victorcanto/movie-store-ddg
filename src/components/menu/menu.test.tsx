import React from 'react';
import { render } from '../../shared/utils/test-utils';
import { Menu } from './menu.component';

test('snapshot of the base state', () => {
	const { asFragment } = render(<Menu />);
	expect(asFragment()).toMatchSnapshot();
});

test('should have in the document', () => {
	const { getByTestId } = render(<Menu />);
	expect(getByTestId('menu-drawer')).toBeInTheDocument();
});
