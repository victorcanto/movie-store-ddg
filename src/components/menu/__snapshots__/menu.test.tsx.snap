// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="MuiDrawer-root MuiDrawer-docked css-7ik032-MuiDrawer-docked"
    data-testid="menu-drawer"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorRight MuiDrawer-paperAnchorDockedRight css-qbio6-MuiPaper-root-MuiDrawer-paper"
      style="visibility: hidden; webkit-transform: translateX(1024px); transform: translateX(1024px);"
    >
      <div
        class="MuiBox-root css-106yi4r"
        role="presentation"
      >
        <div
          class="MuiBox-root css-1b6n4o1"
        >
          <div
            class="MuiBox-root css-399ykn"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
              >
                <div
                  class="MuiBox-root css-0"
                >
                  <div
                    class="MuiBox-root css-69i1ev"
                  >
                    <div
                      class="MuiTypography-root MuiTypography-h4 css-1mtqh3w-MuiTypography-root"
                    />
                    <button
                      aria-label="close"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
                <ul
                  class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
                  data-testid="list-items"
                >
                  <div>
                    <div
                      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
                      style="min-height: 0px;"
                    >
                      <div
                        class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                      >
                        <div
                          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </ul>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-vr9oo4"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
